docs/running-on-kubernetes.md

```diff
@@ -0,0 +1,502 @@
+---
+layout: global
+title: Running Spark on Kubernetes
+---
+* This will become a table of contents (this text will be scraped).
+{:toc}
+
+Spark can run on clusters managed by [Kubernetes](https://kubernetes.io). This feature makes use of native
+Kubernetes scheduler that has been added to Spark.
+
+# Prerequisites
+
+* A runnable distribution of Spark 2.3 or above.
+* A running Kubernetes cluster at version >= 1.6 with access configured to it using
+[kubectl](https://kubernetes.io/docs/user-guide/prereqs/).  If you do not already have a working Kubernetes cluster,
+you may setup a test cluster on your local machine using
+[minikube](https://kubernetes.io/docs/getting-started-guides/minikube/).
+  * We recommend using the latest release of minikube with the DNS addon enabled.
+* You must have appropriate permissions to list, create, edit and delete
+[pods](https://kubernetes.io/docs/user-guide/pods/) in your cluster. You can verify that you can list these resources
+by running `kubectl auth can-i <list|create|edit|delete> pods`.
+  * The service account credentials used by the driver pods must be allowed to create pods, services and configmaps.
+* You must have [Kubernetes DNS](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/) configured in your cluster.
+
+# How it works
+
+<p style="text-align: center;">
+  <img src="img/k8s-cluster-mode.png" title="Spark cluster components" alt="Spark cluster components" />
+</p>
+
+<code>spark-submit</code> can be directly used to submit a Spark application to a Kubernetes cluster.
+The submission mechanism works as follows:
+
+* Spark creates a Spark driver running within a [Kubernetes pod](https://kubernetes.io/docs/concepts/workloads/pods/pod/).
+* The driver creates executors which are also running within Kubernetes pods and connects to them, and executes application code.
+* When the application completes, the executor pods terminate and are cleaned up, but the driver pod persists
+logs and remains in "completed" state in the Kubernetes API until it's eventually garbage collected or manually cleaned up.
+
+Note that in the completed state, the driver pod does *not* use any computational or memory resources.
+
+The driver and executor pod scheduling is handled by Kubernetes. It will be possible to affect Kubernetes scheduling
+decisions for driver and executor pods using advanced primitives like
+[node selectors](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector)
+and [node/pod affinities](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity)
+in a future release.
+
+# Submitting Applications to Kubernetes
+
+## Docker Images
+
+Kubernetes requires users to supply images that can be deployed into containers within pods. The images are built to
+be run in a container runtime environment that Kubernetes supports. Docker is a container runtime environment that is
```

Comment:
> a container runtime environment that Kubernetes supports

Does the current code support these other container runtimes, or just docker? From my memory (and the propery names) it seems that only docker is supported?

Reply:
the images should be runnable by other container runtimes (cri-o, rkt, etc), although I'm not aware of anybody testing that.

Reply:
@vanzin, we don't make any runtime level assumptions in Spark code. The K8s abstraction layer [CRI](http://blog.kubernetes.io/2016/12/container-runtime-interface-cri-in-kubernetes.html) should in theory allow using a differnet runtime.

Reply:
My comment is more about the properties being called "docker" and whether that means only docker images are supported. If you can use any image supported by the k8s cluster, than pehaps the properties should be renamed.

Reply:
I see your point - Although there is flexibility in theory, as of now, it's safe to assume that *most* people are running docker containers when using k8s - making the name docker much more intuitive. If the other runtimes do see traction in future (and we do some testing around them), we can rename to `container.image` instead of `docker.image`. As of now, I can make the documentation clearer that spark on k8s only supports docker images. Sound like a reasonable thing to do here?

Reply:
Just my two cents, if we could foresee that we may add support to other containers, then we should consider using `container.image` instead of `docker.image` at the beginning, because rename a config is also considered a kind of behavior change, that commonly takes more effort to process. If it is not that case (we are satisfied to be only supporting docker containers), then it would be fine to just keep the `docker.image` phrase.

Reply:
I wonder if k8s can tell the container runtime from the image name? If it can, we can use `container.image` here, but otherwise, I guess we need another config like `xxx.container` to tell the container runtime and `xxx.${container}.image` to specify the image, e.g. `xxx.container=docker` and `xxx.docker.image=something`.

Reply:
+1 to @ueshin's comment; if the same property supports more than docker images, it should be renamed to use a more generic name.

Reply:
Makes sense. I think `container.image` should be fine then - will send a PR changing that. 
I don't think there are plans to support choosing the runtime on the fly, or letting the CRI abstraction leak into the application layer.

Reply:
I'd expect the container run-time to be transparent to the particular image, and very definitely transparent to spark.  Using `container.image` makes sense from either standpoint.

Reply:
https://github.com/apache/spark/pull/19995 PTAL